packages/grid/_modules_/grid/hooks/features/rows/useGridRowsHydration.ts

```diff
@@ -0,0 +1,81 @@
+import * as React from 'react';
+import { GridApiRef } from '../../../models/api/gridApiRef';
+import { GridRowsHydrationApi } from '../../../models/api/gridRowsHydrationApi';
+import { DataGridProcessedProps } from '../../../models/props/DataGridProps';
+import { getCurrentPageRows } from '../../utils/useCurrentPageRows';
+import { useGridApiMethod } from '../../utils/useGridApiMethod';
+import { GridRowId } from '../../../models/gridRows';
+import { useGridSelector } from '../../utils/useGridSelector';
+import {
+  gridDensityRowHeightSelector,
+  gridDensityFactorSelector,
+} from '../density/densitySelector';
+import { gridFilterStateSelector } from '../filter/gridFilterSelector';
+import { gridPaginationSelector } from '../pagination/gridPaginationSelector';
+import { gridSortingStateSelector } from '../sorting/gridSortingSelector';
+
+/**
+ * @requires useGridPageSize (method)
+ * @requires useGridPage (method)
+ */
+export const useGridRowsHydration = (
+  apiRef: GridApiRef,
+  props: Pick<DataGridProcessedProps, 'getRowHeight' | 'pagination' | 'paginationMode'>,
+): void => {
+  const { getRowHeight, pagination, paginationMode } = props;
+  const rowsHeightLookup = React.useRef<{ [key: GridRowId]: number }>({});
+  const rowHeight = useGridSelector(apiRef, gridDensityRowHeightSelector);
+  const filterState = useGridSelector(apiRef, gridFilterStateSelector);
+  const paginationState = useGridSelector(apiRef, gridPaginationSelector);
+  const sortingState = useGridSelector(apiRef, gridSortingStateSelector);
+
+  const hydrateRowsMeta = React.useCallback(() => {
+    const { rows } = getCurrentPageRows(apiRef.current.state, {
+      pagination,
+      paginationMode,
+    });
+
+    apiRef.current.setState((state) => {
+      const positions: number[] = [];
+      const totalHeight = rows.reduce((acc: number, row) => {
+        positions.push(acc);
+        let targetRowHeight = gridDensityRowHeightSelector(state);
+
+        if (getRowHeight) {
+          // Default back to base rowHeight if getRowHeight returns null or undefined.
+          targetRowHeight =
+            getRowHeight({ ...row, densityFactor: gridDensityFactorSelector(state) }) ||
+            gridDensityRowHeightSelector(state);
+        }
+
+        rowsHeightLookup.current[row.id] = targetRowHeight;
+
+        return acc + targetRowHeight;
+      }, 0);
+
+      return {
+        ...state,
+        rows: {
+          ...state.rows,
+          meta: { totalHeight, positions },
+        },
+      };
+    });
+    apiRef.current.forceUpdate();
+  }, [apiRef, pagination, paginationMode, getRowHeight]);
+
+  const getTargetRowHeight = (rowId: GridRowId): number =>
+    rowsHeightLookup.current[rowId] || rowHeight;
+
+  // The effect is used to build the rows meta data - totalHeight and positions.
+  // Because of variable row height this is needed for the virtualization
+  React.useEffect(() => {
```

Comment:
It would be nice to be able to update the row height state synchronously instead that in an effect, to avoid flickering if sometimes a render slides between the visible rows update and the rows meta update.
With the naive example below, we might have several computation when we update the visible rows AND the page at the same time for instance. It would be interesting to have an event for `pageRowsChange` I think. As it is not a trivial change, I think we can keep it for later.

```suggestion
  React.useEffect(() => {
    hydrateRowsMeta();
  }, [hydrateRowsMeta]);

  // The code for the api method

  useGridApiEventHandler(apiRef, GridEvents.visibleRowsSet, hydrateRowsMeta);
  useGridApiEventHandler(apiRef, GridEvents.pageChange, hydrateRowsMeta);
  useGridApiEventHandler(apiRef, GridEvents.pageSizeChange, hydrateRowsMeta);
```

Reply:
Agreed as a follow-up. I was looking into a solution like this in the beginning but we need to add events for density and `rowHeight` as well in order for this to work. The difficulty would be to add the events emit on all the relevant places and not miss a case.

Reply:
I feel like at one point we may need a hooks that runs a state updater when a selector value has changed.

Something like:

```tsx
export const useGridDerivedState = <V>(
  apiRef: GridApiRef,
  selector: (state: GridState) => V, 
  updater: (state: GridState) => GridState,
) => {
  const prevValueRef = React.useRef<V>();
  
  useFirstRender(() => {
    prevValueRef.current = selector(apiRef.current.state);
  });

  const handleStateChange = React.useCallback(() => {
    const newValue = selector(apiRef.current.state);
    if (newValue !== prevValueRef.current) {
      prevValueRef.current = newValue;
      apiRef.current.setState(updater);
    }
  }, [apiRef]);

  useGridApiEventHandler(apiRef, GridEvents.stateChange, handleStateChange);
}
```

Or even with a single `setState` with some deeper change.